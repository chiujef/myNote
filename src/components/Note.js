import PropTypes from 'prop-types';
import React, {Fragment} from 'react';
import {ScrollView, StyleSheet, TextInput, View} from 'react-native';
import {Button} from 'react-native-elements';
import {ControlHeader} from '@components/core';
import {NoteItems} from '@components';
import {NoteItemsContainer} from '@containers';

export default class Note extends React.Component {
  constructor(props) {
    super(props);

    this.noteItemsContainer = React.createRef();
  }

  onNoteItemAddPress = () => {
    this.noteItemsContainer.current.onNoteItemAddPress();
  };

  render() {
    const {note, onTitleChange, onNoteItemChange} = this.props;

    return (
      <Fragment>
        <ControlHeader>
          <Button title="+ Add" onPress={this.onNoteItemAddPress} />
        </ControlHeader>
        <ScrollView
          style={styles.noteContainer}
          contentInsetAdjustmentBehavior="automatic"
          keyboardShouldPersistTaps="handled">
          <TextInput
            onChangeText={onTitleChange}
            style={styles.noteTitle}
            maxLength={25}
            placeholder="title"
            value={note.title}
          />
          <View>
            <NoteItemsContainer
              ref={this.noteItemsContainer}
              noteItems={note.items}
              onNoteItemChange={onNoteItemChange}>
              <NoteItems />
            </NoteItemsContainer>
          </View>
        </ScrollView>
      </Fragment>
    );
  }
}

const styles = StyleSheet.create({
  noteContainer: {
    marginTop: 32,
    paddingHorizontal: 24,
  },
  noteTitle: {
    fontSize: 24,
    fontWeight: '600',
  },
});

Note.propTypes = {
  note: PropTypes.shape({
    title: PropTypes.string,
    items: PropTypes.arrayOf(
      PropTypes.shape({
        id: PropTypes.number,
        text: PropTypes.string,
        completed: PropTypes.bool,
      }),
    ),
  }),
  onTitleChange: PropTypes.func,
  onNoteItemChange: PropTypes.func,
};
