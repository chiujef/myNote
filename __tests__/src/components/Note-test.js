import 'react-native';
import React from 'react';
import renderer from 'react-test-renderer';
import {render, fireEvent} from 'react-native-testing-library';

import Note from '@components/Note';

describe('Note', () => {
  let note;

  beforeEach(() => {
    note = {
      title: 'this is the title',
      items: [
        {id: 1, text: 'Text lorem ipsum something', completed: false},
        {id: 2, text: 'Text1', completed: false},
        {
          id: 3,
          text:
            'Text2 what is this? Lorem ipsum lang akong nahibaw an. la nay lain.',
          completed: true,
        },
      ],
    };
  });

  describe('Rendering', () => {
    it('should match to snapshot', () => {
      const component = renderer.create(<Note note={note} />).toJSON();
      expect(component).toMatchSnapshot();
    });

    it('should display title', () => {
      const {getByDisplayValue} = render(<Note note={note} />);

      expect(getByDisplayValue(note.title)).not.toBeNull();
    });

    it('should display the note item texts', () => {
      const {getAllByText} = render(<Note note={note} />);

      expect(getAllByText(/Text/).length).toBe(3);
    });
  });

  describe('Changing title', () => {
    it('should trigger onTitleChange with new value', () => {
      const onTitleChangeHandler = jest.fn();
      const {getByDisplayValue} = render(
        <Note note={note} onTitleChange={onTitleChangeHandler} />,
      );

      fireEvent.changeText(getByDisplayValue(note.title), 'ab');

      expect(onTitleChangeHandler).toHaveBeenCalledWith('ab');
    });
  });

  describe('Pressing "Add"', () => {
    it('should trigger onNoteItemChange event with new noteItems values', () => {
      const onNoteItemChangeHandler = jest.fn();
      const {getByText} = render(
        <Note note={note} onNoteItemChange={onNoteItemChangeHandler} />,
      );

      fireEvent.press(getByText('+ Add'));

      expect(onNoteItemChangeHandler.mock.calls[0][0].length).toBe(4);
    });
  });
});
