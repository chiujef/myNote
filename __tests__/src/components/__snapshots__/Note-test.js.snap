// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Note Rendering should match to snapshot 1`] = `
Array [
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#F5FCFF",
        "borderTopColor": "#b3bec9",
        "borderTopWidth": 1,
        "flexDirection": "row",
        "height": 50,
        "justifyContent": "center",
        "position": "absolute",
        "top": 0,
        "width": "100%",
      }
    }
  >
    <View
      style={
        Object {
          "borderRadius": 3,
        }
      }
    >
      <View
        accessible={true}
        clickable={true}
        isTVSelectable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "#2089dc",
              "borderColor": "#2089dc",
              "borderRadius": 3,
              "borderWidth": 0,
              "flexDirection": "row",
              "justifyContent": "center",
              "padding": 8,
            }
          }
        >
          <Text
            style={
              Object {
                "color": "white",
                "fontSize": 18,
                "paddingBottom": 1,
                "paddingTop": 2,
                "textAlign": "center",
              }
            }
          >
            + Add
          </Text>
        </View>
      </View>
    </View>
  </View>,
  <RCTScrollView
    contentInsetAdjustmentBehavior="automatic"
    keyboardShouldPersistTaps="handled"
    style={
      Object {
        "marginTop": 32,
        "paddingHorizontal": 24,
      }
    }
  >
    <View>
      <TextInput
        allowFontScaling={true}
        maxLength={25}
        placeholder="title"
        rejectResponderTermination={true}
        style={
          Object {
            "fontSize": 24,
            "fontWeight": "600",
          }
        }
        underlineColorAndroid="transparent"
        value="this is the title"
      />
      <View>
        <View
          style={
            Object {
              "flexDirection": "row",
              "justifyContent": "flex-start",
            }
          }
        >
          <View
            accessible={true}
            clickable={true}
            isTVSelectable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "margin": 5,
                "marginLeft": 10,
                "marginRight": 10,
                "opacity": 1,
                "padding": 10,
              }
            }
            testID="checkbox"
          >
            <View
              style={
                Object {
                  "alignItems": "center",
                  "flexDirection": "row",
                }
              }
            >
              <Text
                allowFontScaling={false}
                style={
                  Array [
                    Object {
                      "color": "#bfbfbf",
                      "fontSize": 24,
                    },
                    Object {
                      "minWidth": 24,
                    },
                    Object {
                      "fontFamily": "FontAwesome",
                      "fontStyle": "normal",
                      "fontWeight": "normal",
                    },
                    Object {},
                  ]
                }
              >
                
              </Text>
            </View>
          </View>
          <View
            style={
              Object {
                "flex": 1,
              }
            }
          >
            <Text
              accessible={true}
              clickable={true}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={null}
            >
              Text lorem ipsum something
            </Text>
          </View>
          <View
            style={
              Object {
                "alignItems": "flex-start",
                "flex": 0.25,
                "justifyContent": "center",
                "margin": 1,
              }
            }
          />
        </View>
        <View
          style={
            Object {
              "flexDirection": "row",
              "justifyContent": "flex-start",
            }
          }
        >
          <View
            accessible={true}
            clickable={true}
            isTVSelectable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "margin": 5,
                "marginLeft": 10,
                "marginRight": 10,
                "opacity": 1,
                "padding": 10,
              }
            }
            testID="checkbox"
          >
            <View
              style={
                Object {
                  "alignItems": "center",
                  "flexDirection": "row",
                }
              }
            >
              <Text
                allowFontScaling={false}
                style={
                  Array [
                    Object {
                      "color": "#bfbfbf",
                      "fontSize": 24,
                    },
                    Object {
                      "minWidth": 24,
                    },
                    Object {
                      "fontFamily": "FontAwesome",
                      "fontStyle": "normal",
                      "fontWeight": "normal",
                    },
                    Object {},
                  ]
                }
              >
                
              </Text>
            </View>
          </View>
          <View
            style={
              Object {
                "flex": 1,
              }
            }
          >
            <Text
              accessible={true}
              clickable={true}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={null}
            >
              Text1
            </Text>
          </View>
          <View
            style={
              Object {
                "alignItems": "flex-start",
                "flex": 0.25,
                "justifyContent": "center",
                "margin": 1,
              }
            }
          />
        </View>
        <View
          style={
            Object {
              "flexDirection": "row",
              "justifyContent": "flex-start",
            }
          }
        >
          <View
            accessible={true}
            clickable={true}
            isTVSelectable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "margin": 5,
                "marginLeft": 10,
                "marginRight": 10,
                "opacity": 1,
                "padding": 10,
              }
            }
            testID="checkbox"
          >
            <View
              style={
                Object {
                  "alignItems": "center",
                  "flexDirection": "row",
                }
              }
            >
              <Text
                allowFontScaling={false}
                style={
                  Array [
                    Object {
                      "color": "#2089dc",
                      "fontSize": 24,
                    },
                    Object {
                      "minWidth": 24,
                    },
                    Object {
                      "fontFamily": "FontAwesome",
                      "fontStyle": "normal",
                      "fontWeight": "normal",
                    },
                    Object {},
                  ]
                }
              >
                
              </Text>
            </View>
          </View>
          <View
            style={
              Object {
                "flex": 1,
              }
            }
          >
            <Text
              accessible={true}
              clickable={true}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Object {
                  "textDecorationLine": "line-through",
                  "textDecorationStyle": "solid",
                }
              }
            >
              Text2 what is this? Lorem ipsum lang akong nahibaw an. la nay lain.
            </Text>
          </View>
          <View
            style={
              Object {
                "alignItems": "flex-start",
                "flex": 0.25,
                "justifyContent": "center",
                "margin": 1,
              }
            }
          />
        </View>
      </View>
    </View>
  </RCTScrollView>,
]
`;
